### PR TITLE
Tony sandbox

### PR DESCRIPTION
**Pull request by tginart**

### CODE REVIEW

File: quantizers/uniform.py [hunk 1]
```diff
@@ -59,9 +64,26 @@ def quantize(self, X):
...
+
+    def name(self):
+        return "uniform_mt" + str(self.num_bits) + "b"
+
+    def quantize(self, X):
+        if(self.num_bits <= 1): #mid-tread requires at least 2 bits
+            return np.zeros(X.shape), UniformQuantizer.get_total_bytes(self, X)
+
+        L = 2**self.num_bits-1
+        eps = 1e-7
```
Comment (chrisaberger) [inline quantizers/uniform.py]: What is this?

File: quantizers/kmeans.py [hunk 2]
```diff
@@ -12,7 +12,7 @@ def name(self):
         return "kmeans" + str(self.num_centroids) + "b"
 
     def quantize(self, X):
-        kmeans = KMeans(n_clusters=self.num_centroids)\
+        kmeans = KMeans(n_clusters=self.num_centroids,max_iter=120,tol=0.01)\
```
Comment (chrisaberger) [inline quantizers/kmeans.py]: Can you run yapf on this formatting?

File: quantizers/uniform.py [hunk 3]
```diff
@@ -59,9 +63,27 @@ def quantize(self, X):
...
+
+class MidtreadQuantizer(UniformQuantizer):
+    def name(self):
+        return "uniform_mt" + str(self.num_bits) + "b"
+
+    def quantize(self, X):
+        if (self.num_bits <= 1):  #mid-tread requires at least 2 bits
+            return np.zeros(X.shape), UniformQuantizer.get_total_bytes(self, X)
+
+        L = 2**self.num_bits - 1
```
Comment (chrisaberger) [inline quantizers/uniform.py]: shouldn't this be 2**(self.num_bits-1)?

File: quantizers/uniform.py [hunk 4]
```diff
@@ -59,9 +63,27 @@ def quantize(self, X):
...
+        eps = 1e-7
+        a = max(np.max(X), -1 * np.min(X)) + eps
```
Comment (chrisaberger) [inline quantizers/uniform.py]: What if np.min is a positive number? Shouldn't "-1 * np.min(X)" be "np.abs(np.min(X))"?

Comment (chrisaberger) [reply quantizers/uniform.py]: Same with np.max? What if it is a negative number?

File: quantizers/uniform.py (hunk 3 above)
Comment (chrisaberger) [reply quantizers/uniform.py]: For clarity please change this to "(2**(self.num_bits-1))-1" please add a comment saying this is the maximum positive number you can represent.

File: quantizers/uniform.py [hunk 5]
```diff
@@ -59,9 +63,27 @@ def quantize(self, X):
...
+        delta = 2 * a / (L - 1)
```
Comment (chrisaberger) [inline quantizers/uniform.py]: Delete the L-1 and just do it when you declare L above. Now if a is the negative number this is not 100% correct though. If a is the negative number there should be no -1 on the L. You should comment your code...

File: quantizers/uniform.py (hunk 4 above)
Comment (tginart) [reply quantizers/uniform.py]: FIX: change to abs

**Pull request merged**